[RNMobile][Embed block] Alignment options

fluiddot opened the pull request:
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3787

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Enable alignment options and implement block styling based on the selected option.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### Default alignment
1. Add an Embed block
2. Add a valid embed URL
3. Observe that the preview content is centered and covers the entire block

### Align left
1. Add an Embed block
2. Add a valid embed URL
3. Tap on the align options located in the toolbar
4. Tap on "Align left"
5. Observe that the preview content is aligned left and only covers part of the block

### Align center
1. Add an Embed block
2. Add a valid embed URL
3. Tap on the align options located in the toolbar
4. Tap on "Align center"
5. Observe that the preview content is centered and covers the entire block

### Align right
1. Add an Embed block
2. Add a valid embed URL
3. Tap on the align options located in the toolbar
4. Tap on "Align right"
5. Observe that the preview content is aligned right and only covers part of the block

### Wide width
1. Add an Embed block
2. Add a valid embed URL
3. Tap on the align options located in the toolbar
4. Tap on "Wide width"
5. Observe that the block's width is wider than the rest of the content and that the preview content covers the entire block

**NOTE:** Some embed content like Tweets won't cover the entire block because it's not designed for that purpose.

### Full width
1. Add an Embed block
2. Add a valid embed URL
3. Tap on the align options located in the toolbar
4. Tap on "Full width"
5. Observe that the block's width covers the entire editor and that the preview content covers the entire block

**NOTE:** Some embed content like Tweets won't cover the entire block because it's not designed for that purpose.

## Screenshots <!-- if applicable -->

<details><summary><strong>Empty align</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="435" alt="align-empty-resize-native" src="https://user-images.githubusercontent.com/14905380/129926919-8a6a6fda-82ec-4ce6-93f8-ba2644f974ab.png">|<img width="817" alt="align-empty-resize-web" src="https://user-images.githubusercontent.com/14905380/129926934-44bd11b3-057e-49ea-b7d0-dde66947d781.png">|<img width="645" alt="align-empty-resize-preview" src="https://user-images.githubusercontent.com/14905380/129926944-b874271c-1219-4a67-aac7-522a237b1407.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="438" alt="align-empty-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129927067-3ef74173-b8b7-4913-b925-d981557dd124.png">|<img width="816" alt="align-empty-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129927079-6e1d3d99-b827-4670-98a1-f3feced2a6ba.png">|<img width="643" alt="align-empty-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927086-6e85f194-77c6-4b3e-97db-5903e1e7b805.png">

</details>

<details><summary><strong>Align left</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="439" alt="align-left-resize-native" src="https://user-images.githubusercontent.com/14905380/129927325-69594b69-3072-49e7-a5db-68b9648e7af3.png">|<img width="796" alt="align-left-resize-web" src="https://user-images.githubusercontent.com/14905380/129927335-94c87703-0780-43f3-bfb4-d714eee431ba.png">|<img width="630" alt="align-left-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927342-3b833c40-c41e-409d-9829-c9502438b06f.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="440" alt="align-left-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129927356-7fc2be60-ace4-422e-a1a8-ae31dd1ebdf9.png">|<img width="819" alt="align-left-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129927360-42af644c-d8ed-4ceb-a245-25c2bbeab7c5.png">|<img width="630" alt="align-left-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927364-3d1a1d93-7f7c-4774-811c-bd6635f2ecbe.png">

</details>

<details><summary><strong>Align right</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="434" alt="align-right-resize-native" src="https://user-images.githubusercontent.com/14905380/129927534-a9b19102-a84c-4313-89fc-d230747a8706.png">|<img width="814" alt="align-right-resize-web" src="https://user-images.githubusercontent.com/14905380/129927545-08fa735e-1a72-4a82-bd9f-409b74b790de.png">|<img width="640" alt="align-right-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927551-57ebdb73-936c-4bec-b776-36586e05951d.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="442" alt="align-right-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129927566-87f7e447-f4a0-4cf6-9074-0d27b7192a30.png">|<img width="818" alt="align-right-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129927569-48b33d5a-a561-4d56-8045-e96f2be05d47.png">|<img width="640" alt="align-right-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927572-0fa2bbcf-c135-4824-b36b-75c239d51688.png">

</details>

<details><summary><strong>Align center</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="439" alt="align-center-resize-native" src="https://user-images.githubusercontent.com/14905380/129927613-af8b19e0-ee8a-44e3-90d6-f30c48f56fda.png">|<img width="808" alt="align-center-resize-web" src="https://user-images.githubusercontent.com/14905380/129927622-db1a9446-ee83-49ee-89e8-0754270dc60e.png">|<img width="634" alt="align-center-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927628-5f4c2f2e-82bf-473c-af82-91c8f969ab84.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="439" alt="align-center-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129927650-2e9c6f8d-3e47-4a46-ada0-650e0b6c8f4f.png">|<img width="809" alt="align-center-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129927656-78a18b82-fb29-4181-a6cd-e223f64e23a4.png">|<img width="635" alt="align-center-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927667-b4e3afaf-fcf4-474b-8e9a-79dff073277b.png">

</details>

<details><summary><strong>Wide width</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="742" alt="wide-width-resize-native" src="https://user-images.githubusercontent.com/14905380/129927759-4442384b-aa67-4e1b-8c3c-e6e2075883ac.png">|<img width="1067" alt="wide-width-resize-web" src="https://user-images.githubusercontent.com/14905380/129927764-581596b0-6320-4967-a4f7-977e58ffb5a1.png">|<img width="904" alt="wide-width-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927767-1d91e2d3-77c9-42af-93d6-3c859f8e99df.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="745" alt="wide-width-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129927789-b3bf02c3-c373-4883-a7f6-be832d5e6968.png">|<img width="1060" alt="wide-width-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129927810-8a92ba7b-2e31-461f-b9fa-b1e3c878b292.png">|<img width="709" alt="wide-width-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129927817-a7035b18-e6f5-48db-be78-30536c79b4c6.png">

</details>

<details><summary><strong>Full width</strong></summary>

#### Resize mode ON

Native editor|Web editor|Tablet preview
--|--|--
<img width="770" alt="full-width-resize-native" src="https://user-images.githubusercontent.com/14905380/129928311-d1130c3c-05c8-4634-904c-59e7f71e4ed1.png">|<img width="1331" alt="full-width-resize-web" src="https://user-images.githubusercontent.com/14905380/129928322-dd75017e-2d5c-4d8c-9a75-90302cc00e57.png">|<img width="941" alt="full-width-resize-preview" src="https://user-images.githubusercontent.com/14905380/129928327-ac40fb2b-8f88-42fe-b7d0-90d09b4ec52f.png">

#### Resize mode OFF

Native editor|Web editor|Tablet preview
--|--|--
<img width="760" alt="full-width-no-resize-native" src="https://user-images.githubusercontent.com/14905380/129928356-bd14f624-b100-4ecf-b9c0-b9bbd5711e56.png">|<img width="1290" alt="full-width-no-resize-web" src="https://user-images.githubusercontent.com/14905380/129928373-5045c28e-5d27-443b-abe2-7a07fcc3ded1.png">|<img width="712" alt="full-width-no-resize-preview" src="https://user-images.githubusercontent.com/14905380/129928378-658de132-b5d0-466c-98cb-536ddfbff252.png">

</details>

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jd-alexander committed:
Created an EmbedPreview component to house the caption.

jd-alexander committed:
Disabled the placeholder mode temporarily until the linking is enabled.

jd-alexander committed:
added the react-native-web-view package.

jd-alexander committed:
linked react-native-web-view to ios

jd-alexander committed:
integrated react-native-web-view with android

jd-alexander committed:
initial integration of WebView into the embed-preview component.

jd-alexander committed:
added project configurations for react-native-webview android

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-preview

# Conflicts:
#	package-lock.json
#	packages/block-library/src/embed/edit.js
#	packages/block-library/src/embed/embed-preview.native.js
#	packages/react-native-bridge/android/react-native-bridge/build.gradle
#	packages/react-native-editor/ios/Podfile.lock
#	packages/react-native-editor/package.json

fluiddot committed:
Add sandbox component native version

fluiddot committed:
Use webview to preview embed block

fluiddot committed:
Update aspect ratio calculation in sandbox component

fluiddot committed:
Add native styles to sandbox component

fluiddot committed:
Open embed content in external browser

fluiddot committed:
Add comment to sandbox webview

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-preview

# Conflicts:
#	package-lock.json
#	packages/block-library/src/embed/edit.native.js
#	packages/block-library/src/embed/embed-preview.native.js
#	packages/react-native-editor/package.json

fluiddot committed:
Update react-native-webview package to wp-fork

fluiddot committed:
Bump react-native-webview version

fluiddot committed:
Bump react-native-webview version

fluiddot committed:
Use provider url as base url for embed webview

fluiddot committed:
Add forceRender param to trySandbox

fluiddot committed:
Handle case of setting empty URL after preview

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-preview

# Conflicts:
#	packages/react-native-bridge/android/settings.gradle

fluiddot committed:
Enable embed preview interaction when selected

fluiddot committed:
Add max allowed requests to embed preview

fluiddot committed:
Prevent navigation in embed preview on Android

fluiddot committed:
Recreate WebView in Android on orientation change

fluiddot committed:
Disable embed preview interaction

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-preview

fluiddot committed:
Remove unused clientId prop from embed web version

fluiddot committed:
Enable inline preview only in dev mode

fluiddot committed:
Remove URL from embed block internal state

I noticed that for the mobile version we don't really need to keep the URL in the internal state, as we were only modifying it once it changed, and that's already being handled with the url block's attribute.

fluiddot committed:
Use memo in embed preview component

fluiddot committed:
Use memo in sandbox component

fluiddot committed:
Rename iframe html to content html and update logic

fluiddot committed:
Bump react-native-webview package

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-preview

fluiddot committed:
Add align styles to embed preview

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-alignment

# Conflicts:
#	packages/block-library/src/embed/edit.native.js
#	packages/block-library/src/embed/embed-preview.native.js
#	packages/components/src/sandbox/index.native.js
#	packages/components/src/sandbox/style.native.scss

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-alignment

fluiddot committed:
Add embed aspect ratio styles to sandbox static html

fluiddot committed:
Remove width calculation from sandbox

We really only need to calculate the height as the width should fit the block's size.

fluiddot commented on packages/components/src/sandbox/index.native.js:
```diff
@@ -92,22 +92,62 @@ const style = `
+		overflow: hidden; /* If it has an aspect ratio, it shouldn't scroll. */
+	}
+
+	/* Add responsiveness to embeds with aspect ratios. */
+	.wp-has-aspect-ratio.wp-block-embed__wrapper::before {
+		content: "";
+		display: block;
+		padding-top: 50%; // Default to 2:1 aspect ratio.
+	}
+	.wp-has-aspect-ratio iframe {
+		position: absolute;
+		top: 0;
+		right: 0;
+		bottom: 0;
+		left: 0;
+		height: 100%;
+		width: 100%;
+	}
+	.wp-embed-aspect-21-9.wp-block-embed__wrapper::before {
+		padding-top: 42.85%; // 9 / 21 * 100
+	}
+	.wp-embed-aspect-18-9.wp-block-embed__wrapper::before {
+		padding-top: 50%; // 9 / 18 * 100
+	}
+	.wp-embed-aspect-16-9.wp-block-embed__wrapper::before {
+		padding-top: 56.25%; // 9 / 16 * 100
+	}
+	.wp-embed-aspect-4-3.wp-block-embed__wrapper::before {
+		padding-top: 75%; // 3 / 4 * 100
+	}
+	.wp-embed-aspect-1-1.wp-block-embed__wrapper::before {
+		padding-top: 100%; // 1 / 1 * 100
+	}
+	.wp-embed-aspect-9-16.wp-block-embed__wrapper::before {
+		padding-top: 177.77%; // 16 / 9 * 100
+	}
+	.wp-embed-aspect-1-2.wp-block-embed__wrapper::before {
+		padding-top: 200%; // 2 / 1 * 100
+	}
```
In the web version, these styles are applied in a wrapper element to the sandbox component.
https://github.com/WordPress/gutenberg/blob/253d9b6e2123a2806b8843562f3f37de186c933a/packages/block-library/src/embed/embed-preview.js#L114-L118
https://github.com/WordPress/gutenberg/blob/253d9b6e2123a2806b8843562f3f37de186c933a/packages/block-library/src/embed/embed-preview.js#L95

In the native version, we can't do it the same way because the content is being rendered within a WebView, so for this case, we need to include them in the static HTML.

**Styles in the web version:**
https://github.com/WordPress/gutenberg/blob/253d9b6e2123a2806b8843562f3f37de186c933a/packages/block-library/src/embed/style.scss#L38-L89

ceyhun replied:
Could there be a way to import the related parts from `embed/style.scss` to JS code here, instead of duplicating? If not, maybe we can add a comment mentioning that it is a copy of the styles in `embed/style.scss`?

fluiddot replied:
Not sure if there would be a way, if so I think it could be complex because the styles are converted to the React Native format, which is different from the web so we would need to convert it back to CSS to be used here.

I don't like the idea to have different copies of the same code for both versions but for this case, since the embed previews are rendered in a different way due to the fact that on the native version they're rendered within WebViews, unfortunately, we'll have to maintain a copy.

For now, I'll add a comment to keep a reference to the original styles code 👍 .

fluiddot commented on packages/components/src/sandbox/index.native.js:
```diff
@@ -117,7 +157,6 @@ function Sandbox( {
 	url,
 } ) {
 	const ref = useRef();
-	const [ width, setWidth ] = useState( 0 );
```
We really only need to calculate the height as the width should fit the block's size, so I removed the width calculation.

fluiddot commented on packages/components/src/sandbox/index.native.js:
```diff
@@ -142,7 +181,7 @@ function Sandbox( {
 		// Scripts go into the body rather than the head, to support embedded content such as Instagram
 		// that expect the scripts to be part of the body.
 		const htmlDoc = (
-			<html lang={ lang } className={ type }>
+			<html lang={ lang }>
```
I realized that adding `type` to the `html` tag breaks the styles for the `
Resize for smaller devices` option, besides, I think this is not really necessary as we're already adding it in the `body` tag.

fluiddot commented on packages/block-library/src/embed/styles.native.scss:
```diff
@@ -72,6 +72,24 @@
+	align-items: flex-end;
+}
+
+.embed-preview__sandbox--align-left {
+	max-width: 360px;
+}
+.embed-preview__sandbox--align-right {
+	max-width: 360px;
+}
```
The `max-width` value is based on the styles from the web version:
https://github.com/WordPress/gutenberg/blob/253d9b6e2123a2806b8843562f3f37de186c933a/packages/block-library/src/embed/style.scss#L7

fluiddot commented on packages/block-library/src/embed/embed-preview.native.js:
```diff
@@ -91,13 +101,17 @@ const EmbedPreview = ( {
 							type={ sandboxClassnames }
 							providerUrl={ providerUrl }
 							url={ url }
+							containerStyle={ sandboxAlignStyle }
```
This new prop allows us to control the styles of the WebView component, for this case, we're passing the styles required to align the content.

fluiddot committed:
Add webview content style in sandbox

fluiddot committed:
Disable scroll and zoom features in sandbox

fluiddot committed:
Use html for rendering embed images

Previously embed images were rendering the thumbnail instead of the original content.

fluiddot commented on packages/block-library/src/embed/embed-preview.native.js:
```diff
@@ -61,7 +70,7 @@ const EmbedPreview = ( {
 	}
 
 	const { provider_url: providerUrl } = preview;
-	const html = 'photo' === type ? getPhotoHtml( preview ) : preview.html;
+	const html = preview.html;
```
I realized that for image type embeds the previous code was using the thumbnail instead of the original content which usually is in low resolution.

ceyhun replied:
I think `getPhotoHtml` is also completely replacing the `html` returned from the backend. Could there be a reason for that? Should we still use `url` instead of `thumbnail_url` on mobile but keep replacing the `html`?

fluiddot replied:
> I think getPhotoHtml is also completely replacing the html returned from the backend.
Yes, it's basically rendering a `img` tag with the image's URL:
```
<p>
  <img src={ imageUrl } alt={ photo.title } width="100%" />
</p>
```

> Could there be a reason for that?

That's a good question, I guess this approach might be related to the fact of providing a controlled HTML with custom styles.

> Should we still use url instead of thumbnail_url on mobile but keep replacing the html?

I recall that originally when I checked a Flickr embed on the web version it was using the full-size URL. However, now I've done another check and I noticed that it's displaying the thumbnail 🤔 . At this stage, I think I'm going to revert this change to match the same behavior as on the web version.

fluiddot commented:
@ceyhun the PR is ready to review, it incorporates styles for the alignment attribute but also addresses different style issues like the one related to the aspect ratios ([comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3823#issuecomment-896976305)).

ceyhun commented on packages/block-library/src/embed/embed-preview.native.js:
```diff
@@ -36,6 +37,14 @@ const EmbedPreview = ( {
+	const wrapperStyle = styles[ 'embed-preview__wrapper' ];
+	const wrapperAlignStyle =
+		styles[ `embed-preview__wrapper--align-${ align }` ] ||
+		styles[ 'embed-preview__wrapper--align' ];
```
I couldn't see a `embed-preview__wrapper--align` object in `styles.native.scss`, should we maybe add it or maybe we can remove this fallback?

fluiddot replied:
Good point, originally I defined the style `embed-preview__wrapper--align` but in the last commits, I removed it because it wasn't necessary, so we can safely remove the fallback ✂️ 😄 .

fluiddot replied:
This has been addressed in [this commit](a88818615e6db3514a494ed60a33fb902c412589).

ceyhun commented on packages/block-library/src/embed/embed-preview.native.js:
```diff
@@ -36,6 +37,14 @@ const EmbedPreview = ( {
+		styles[ 'embed-preview__wrapper--align' ];
+	const sandboxAlignStyle =
+		styles[ `embed-preview__sandbox--align-${ align }` ] ||
+		styles[ 'embed-preview__sandbox--align' ];
```
Same for `embed-preview__sandbox--align`

fluiddot replied:
Following with [this comment](https://github.com/WordPress/gutenberg/pull/33851#discussion_r700203113), I'll remove this fallback as it's not required either.

fluiddot replied:
This has been addressed in [this commit](a88818615e6db3514a494ed60a33fb902c412589).

ceyhun reviewed (commented):
I've tested this using both WPAndroid and WPiOS and it's looking great! The code in HTML mode for align changes also look consistent compared to the web editor. Just have a couple of questions.

fluiddot committed:
Remove unused embed style fallbacks

fluiddot committed:
Revert "Use html for rendering embed images"

This reverts commit 34f16363802ae3177747c621f0f99472240cfb10.

fluiddot committed:
Add comment to sandbox styles

It references the styles from the web version that have been adapted for native version.

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-alignment

fluiddot commented:
@ceyhun I've pushed the changes to address the feedback from [your review](https://github.com/WordPress/gutenberg/pull/33851#pullrequestreview-743894352), so it's ready for a second review, thanks for the help 🙇 !

ceyhun reviewed (approved):
:shipit:
